### PR TITLE
Add extensions feature from Julia 1.9 with backward compatibility

### PR DESCRIPTION
**Pull request by b-fg**
This allows the conditional loading of the CUDA, AMDGPU, and WriteVTK packages. Code directly depending on these packages is now located in `ext/`. For example, now the `L₂` for `CuArray` is only loaded when `using CUDA` is called in addition to `using WaterLily`.

This is currently a first implementation without propers checks. To do:
 - [x] Test actual backward compatibility: Tested on 1.8.5 and 1.9.4
 - [x] Figure out why having the optional packages in the main Julia installation (ie. release) still messes up the initialization
 - [x] Check CI tests compat and implement do-nothing `@allowscalar` version so that the CUDA extension can actually extend it: now using `GPUArrays.@allowscalar` which works for both CUDA and AMD.

### CODE REVIEW

**Comment by b-fg** [comment]
This is looking better and close to finishing. Some comments:
 - CI now only triggers when a new (normal, not draft) PR is created and when a push is done into master, such as a merge from a PR. For new draft PRs, CI is not triggered. Within a PR (normal or draft) new pushes do not trigger the CI - triggering can always be done manually if necessary. This should reduce notifications spam of CI failing and reduce the number of tests that we actually launch (important when migrating to a GPU-enabled CI system).
 
 - The new extensions packaging can make the developer workflow a little bit different. Before this, I used to just to develop within the `WaterLily.jl` environment (`cd WaterLily.jl; julia --project; include(...)`) and test new features there. Now, if new features make use of a weak dependency such as `CUDA` or `WriteVTK`, testing code within the `WaterLily.jl` environment is problematic since we cannot do `using CUDA` in there, as the package manager detects that `CUDA` is no within `[deps]`  (it's in `[weakdeps]`) and asks to install such package. Instead, I create a new project and environment called `WaterLilyTests.jl`, I `add WaterLily`, mark it as development package `using Pkg; Pkg.develop(PackageSpec(path="<WaterLily.jl path>"));`, and also add whatever packages I need to test development such as CUDA: `add CUDA`. When `WaterLily` detects that CUDA is loaded (`using CUDA`) it will compile the code  in `ext/WaterLilyCUDAExt.jl` and make the extensions available.

**Comment by b-fg** [comment]
So everything is pretty much done. I am locally running the test suite and only current issue is that tests are correctly passing for 1.8.5 and 1.9.4, but no 1.10.0. Specifically, this test is failing on 1.10.0:
https://github.com/weymouth/WaterLily.jl/blob/fe9827338df7c6e53bfefb749fe32d5a0cf93e74/test/runtests.jl#L298
Any idea why this might be happening @weymouth?

**Comment by b-fg** [comment]
Huh, I just did a merge from master and now is throwing some NaNs. Adding the `verbose`  flag in the test I see:
```
tU/L=0.0312, Δt=0.355
tU/L=0.0756, Δt=NaN
tU/L=NaN, Δt=NaN
```

**Comment by b-fg** [comment]
Seems like `exitBC=true` is needed in that test so that the simulation does not blow up in 1.10.0. But then again the other tests after this one which use `exitBC=false` for `Array` and `CuArray` work fine... So for the moment I will modify this test and maybe leave compat to 1.9 only until this is figured out.

**Comment by weymouth** [comment]
That's weird. Also notice that it's failing on macOS again, even for 1.9. We must have something not quite right in that test.

**Comment by b-fg** [comment]
Yeap!

**Comment by weymouth** [comment]
I'll try it on my home laptop, which is a mac.

**Comment by b-fg** [comment]
Did you have a chance to look at this @weymouth?

**Comment by weymouth** [comment]
Nope, but I'll check this weekend.

**Comment by weymouth** [comment]
I can't replicate this error on my Mac with julia 1.10 or 1.09. I think this must be a git CI thing. It's a really simple test which is being violated, which has nothing to do with the `exitBC` state and it only fails sporadically and only online. How are we supposed to debug something like this??

**Comment by weymouth** [comment]
> Seems like `exitBC=true` is needed in that test so that the simulation does not blow up in 1.10.0. But then again the other tests after this one which use `exitBC=false` for `Array` and `CuArray` work fine... So for the moment I will modify this test and maybe we could compat to 1.9 only until this is figured out.

I just tried both 1.9 and 1.10 on Mac and they worked fine. I think we should report this to github and either comment out the test or not worry too much about it failing.

**Comment by weymouth** [comment]
When I run `] test` on this PR, the CUDA and AMDGPU both throw errors when they try to load (since this laptop doesn't have a GPU) but the tests still run and pass. Is that the correct behaviour?

**Comment by b-fg** [comment]
Yes, that is expected since we have it hardcoded in tests hence it tries to load those. Otherwise it would never try to pass them for GPUs, even if CUDA or AMDGPU was functional. I think there is no better way to this at the moment. The main benefit is that for a normal usage (outside of tests) users now do not download and install CUDA/AMDGPU by default.

File: Project.toml [hunk 1]
```diff
@@ -49,4 +49,4 @@ UnicodePlots = "b8865327-cd53-5732-bb35-84acbb429228"
 WriteVTK = "64499a7a-5c06-52f2-abe2-ccb03c286192"
 
 [targets]
-test = ["AMDGPU", "BenchmarkTools", "CUDA", "GLMakie", "GPUArrays", "LaTeXStrings", "Makie", "PerformanceTestTools", "Plots", "Test", "UnicodePlots"]
+test = ["Test", "CUDA", "AMDGPU", "GPUArrays"]
```
Comment (b-fg) [inline Project.toml:52]: Yes nice, this was actually the last cleanup I needed to do in this PR :) So thanks for doing it!

File: ext/WaterLilyAMDGPUExt.jl [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+__init__() = @assert AMDGPU.functional()
+
+AMDGPU.allowscalar(false) # disallow scalar operations on GPU
+
+"""
+    L₂(a)
+
+L₂ norm of ROCArray `a` excluding ghosts.
+"""
+L₂(a::ROCArray,R::CartesianIndices=inside(a)) = mapreduce(abs2,+,@inbounds(a[R]))
```
Comment (b-fg) [inline ext/WaterLilyAMDGPUExt.jl:26]: GPU tests cannot really be covered in the current Github CI.

Comment (weymouth) [reply ext/WaterLilyAMDGPUExt.jl:26]: That's clear.

**Comment by weymouth** [comment]
This ready?

**Pull request merged**